### DISCUSSION TITLE
separate the backend and frontend projects?

### DISCUSSION THREAD

**Discussion by annieappflowy** [AppFlowy-IO/AppFlowy#12] [Ideas]
by @joaquim :

This product looks really nice, congrats on the launch 🙂
i have one suggestion, maybe separate the backend and frontend projects?
So that you could have different people working on different frontends 🙂 like a web one or maybe an  electron 😛  one 
from what ive seen by looking at the source code it would be really easy since you already have the back end running as a local server 
also, great to see a fellow 🇧🇷 on the project

**Comment by annieappflowy** [comment]
by @nathan.fooo:
Thanks a bunch for your input! That's very thoughtful. I initially wanted to separate the backend and the frontend, but I had a challenge about unit tests. Do you have any suggestions on how to deal with unit tests that test both the backend and the frontend?

**Reply by CodeTripper** [reply]
@annieappflowy for the unit testing the Frontend, I suggest you try contract testing using [Pact](https://github.com/pact-foundation). This would help you in testing the frontend as well as ensuring that the backend never breaks the frontend if the backend apis accidentally introduces some breaking changes.

**Comment by annieappflowy** [comment]
by @joaquim

Hey!
Usually at work we just make the front end use a mock for the backend, but i know its kinda hard to maintain the mocks updated
Since the frontend depends on the backend you could build the backend and run it as a docker machine (or just download a release from the backend repo) when running the CI for the frontend 
or just have another repository alltogether that just builds both and run the integration tests
testing is hard 🥲

**Comment by annieappflowy** [comment]
^ copied from the Discord channel

We'd love to log this and open to more suggestions. We plan to build a collaboration server that enables multiple users edit a page at the same time, so this topic will get more and more relevant as we progress.

**Reply by heyakyra** [reply]
You're referencing #714 and #256 , yes?

**Reply by annieappflowy** [reply]
@heyakyra  yes

**Comment by appflowy** [comment]
FYI, The unit tests contained backend and frontend located in /backend/tests/document/edit.rs.
